### DISCUSSION TITLE
find the 300-triads

### DISCUSSION THREAD

**Discussion by koko12131** [networkx/networkx#6158] [Q&A]
i  want to extrcat the triads-300 in large weighted directed networks with networkx 
you can find all possible kinds of triads below
[16 possible triads](https://stackoverflow.com/questions/55291704/networkx-list-of-directed-edges-in-a-triad-census)

here is my codes 
````
network_data = nx.read_gexf(file_path)
G = nx.DiGraph()
G_contain_triple = nx.triads_by_type(network_data)
for x, y in G_contain_triple.items():
    if x == "300":
        for tt in y:
            print(tt.edges.data("weight"))
            G.add_weighted_edges_from(tt.edges.data("weight"))
if len(G.nodes()) > 0:
    nx.write_gexf(G, os.path.join(output_path, year + 'tri.gexf'))
````

the networkx will find all 16 possible types of triads together with nx.triads_by_type(network_data), this will cost very very very much  time. but what i need is only the 300-triads

i wonder if there is any solution to find only the 300-triads without others and save my time you can find the test file 2021_DiGraph_weighted.gexf below
[test file](https://github.com/koko12131/patentee)

**Comment by dschult** [comment]
Another way to ask this question is: can you find the cliques with 3 nodes? Or can you find the triangles. It seems like you should "just" loop over the nodes, look at the neighbors and see whether any of their neighbors include the original node. A good first step would be to form a new graph that only includes edges that are in both direction on the original graph. That should reduce the size of the problem quite a bit for many graphs.  The code for `traidic_census()` has this kind of thing buried in it with all the checks for other triads.

**Comment by dschult** [comment]
I can't resist a puzzle when I'm distracted... :). Here's some code that collects the `300`-triads as a set of 3-tuples. Note that this will list all permutations of the triples. so the last line removes the duplicates (assuming the nodes are sortable)
```python
dbl_nbrs = {n: G.pred[n].keys() & G.succ[n].keys() for n in G}
tt = {((n,nbr,new) for n in dbl_nbrs for nbr in dbl_nbrs[n] for new in dbl_nbrs[n] & dbl_nbrs[nbr] if new not in (n, nbr)))
triples = {tuple(sorted(t)) for t in tt}
```
